modules/create_timepoint/php/timepoint.class.inc

```diff
@@ -0,0 +1,344 @@
+
+        if (!empty($errors)) {
+            $values['errors'] = $errors;
+        }
+        if (empty($values['psc'])) {
+            unset($values['psc']);
+        }
+
+        return new \LORIS\Http\Response\JsonResponse(
+            $values
+        );
+    }
+    /**
+     * Processes the values & saves to database and return a json response.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handlePOST(ServerRequestInterface $request) : ResponseInterface
+    {
+        // Parse POST request body.
+        $values = json_decode((string) $request->getBody(), true);
+
+        // Validate POST data.
+        $values = $this->_validate($values);
+
+        // User permission verification.
+        $values = $this->_verifyPermissions($request, $values);
+        if (isset($values['error'])) {
+            switch ($values['error']) {
+            case 'BadRequest':
+                return new \LORIS\Http\Response\JSON\BadRequest(
+                    'Failed to retrieve data for create timepoint.'
+                );
+            case 'Forbidden':
+                return new \LORIS\Http\Response\JSON\Forbidden(
+                    'Forbidden'
+                );
+            default:
+                break;
+            }
+        }
+
+        // convert site entered to a \Site object
+        $user = \User::singleton();
+
+        $user_list_of_sites = $user->getCenterIDs();
+        $num_sites          = count($user_list_of_sites);
+        $site = null;
+        if ($num_sites == 1) {
+            $site = \Site::singleton($user_list_of_sites[0]);
+        } else if ($num_sites > 1) {
+            $site = \Site::singleton(intval($values['psc']));
+        }
+
+        // Project
+        $project = null;
+        $user_list_of_projects = $user->getProjectIDs();
+        if (count($user_list_of_projects) == 1) {
+            //if there is only one project, autoselect first project from array of 1
+            $project = \Project::getProjectFromID(array_pop($user_list_of_projects));
+        } else if (count($user_list_of_projects) > 1) {
+            $project_id = intval($values['psc']);
+            $project    = \Project::getProjectFromID($project_id);
+        }
+
+        $errors   = $values['errors'];
+        $response = array();
+        if (!$errors) {
+            \TimePoint::createNew(
+                \Candidate::singleton(new CandID($values['candID'])),
+                intval($values['subproject']),
+                $values['visit'] ?? null,
+                $site,
+                $project,
+                intval($values['languages']) ?? null,
+            );
+            $response['status'] = 'success';
+        } else {
+            $response['status'] = 'error';
+            $response['errors'] = $errors;
+        }
+        return new \LORIS\Http\Response\JsonResponse(
```

Comment:
might be good to return https://github.com/aces/Loris/blob/master/src/Http/Response/JSON/OK.php if status is success or the 201 Created.php so the response is more robust

Reply:
@zaliqarosli I'm a little confused what you mean by that? The browser should be receiving 201 if it successfully returns the json response.

Reply:
ahh then you shouldnt be returning a 201 json response with 'errors'. if there are errors, it shouldn't be a 201

Reply:
@zaliqarosli maybe I should rename errors to something else? The old form logic would show a custom error message depending on specific submission mistakes. I'm actually not sure if they can ever be shown anymore with the frontend change of select. It has been awhile since I've written the code. I'll see if I can just remove it or if it makes sense to have it.

Reply:
@zaliqarosli i'll rename errors to validation

Reply:
actually I'm not happy with that name

Reply:
maybe "formValidation" is better

Reply:
@zaliqarosli I guess the confusion came from how I had the functionality of form messages be assigned to errors. I renamed to message when formMessage exist. The messages are not errors and instead are these examples:
You must be affiliated with a site to create a timepoint.
Site must be selected from the available dropdown.
A language must be selected.

Reply:
hey @maltheism , seems like from what i gathered on the internet, that you could either return  a 200 with json errors, or return a 406 or 422. seeing as we don't have those response classes yet, i think it's okay for us to return a 200/201, and I think your fetch api implementation in the JS class is fine!